Update MUI, React, Typescript

mz8i opened the pull request:
Updates Material UI to v5, introduces minor fixes using newly available features.
Also updates React and Typescript to latest versions (this is required by the MUI update).

The MUI update changes the visual style of some UI elements, most notably the default colours of buttons, the nav bar etc. This can be addressed in a separate set of changes if keeping the previous style is desirable.

tomalrussell commented:
Good to keep on the updates treadmill :+1:

As this affects the lockfile so much, can we update it to v2 format? npm v8 complains and rewrites it when running `npm i` by default.

Header background color has changed - reset to `#3f51b5`

Minor spacing issues:
- sidebar box root should have `padding: 16px 32px 32px 32px;` (`padding-top: 16px;` is now being overridden by the more general `padding: 32px;`)
- map control buttons in top left (search and layers) have `margin-top: 8px;`, should flip to `margin-bottom:8px;` to match zoom control distance from the edge of the map.

mz8i commented:
@tomalrussell I've implemented all the suggestions, thank you. Perhaps worth noting that the change of nav bar colour seems like part of a general palette change across MUI. Other UI elements such as buttons, sliders, switches etc have also changed colour, so if we keep just the navbar's old colour, this introduces a slight dissonance in the colours. The question is whether the nav bar colour is something that we want to stick to?

mz8i commented:
Yes, we can, and that's better in terms of consistency. I switched to setting the theme colour. It seems however that MUI changed which of the theme colours it uses for Nav and Buttons - previously the controls were of a reddish colour so didn't use the primary colour. Now, when setting primary colour to `#3f51b5`, all elements use that. So there's still a change from the original, but because it's already more consistent, I'm going to leave it.
